contracts/contracts/strategies/uniswap/UniswapV3LiquidityManager.sol

```diff
@@ -0,0 +1,962 @@
+     *          to Withdraw specified amount of the given asset.
+     *
+     * @param position  Position object
+     * @param asset    Token needed
+     * @param amount    Minimum amount to liquidate
+     *
+     * @return liquidity    Liquidity to burn
+     * @return minAmount0   Minimum amount0 to expect
+     * @return minAmount1   Minimum amount1 to expect
+     */
+    function _calculateLiquidityToWithdraw(
+        Position memory position,
+        address asset,
+        uint256 amount
+    )
+        internal
+        view
+        returns (
+            uint128 liquidity,
+            uint256 minAmount0,
+            uint256 minAmount1
+        )
+    {
+        (uint160 sqrtRatioX96, , , , , , ) = pool.slot0();
+
+        // Total amount in Liquidity pools
+        (uint256 totalAmount0, uint256 totalAmount1) = helper
+            .getAmountsForLiquidity(
+                sqrtRatioX96,
+                position.sqrtRatioAX96,
+                position.sqrtRatioBX96,
+                position.liquidity
+            );
+
+        if (asset == token0) {
+            minAmount0 = amount;
+            minAmount1 = totalAmount1 / (totalAmount0 / amount);
+            liquidity = helper.getLiquidityForAmounts(
+                sqrtRatioX96,
+                position.sqrtRatioAX96,
+                position.sqrtRatioBX96,
+                amount,
+                minAmount1
+            );
+        } else if (asset == token1) {
+            minAmount0 = totalAmount0 / (totalAmount1 / amount);
+            minAmount1 = amount;
+            liquidity = helper.getLiquidityForAmounts(
+                sqrtRatioX96,
+                position.sqrtRatioAX96,
+                position.sqrtRatioBX96,
+                minAmount0,
+                amount
+            );
+        }
+    }
+
+    /**
+     * @notice Liquidiates active position to remove required amount of give asset
+     * @dev Doesn't have non-Reentrant modifier since it's supposed to be delegatecalled
+     *      only from `UniswapV3Strategy.withdraw` which already has a nonReentrant check
+     *      and the storage is shared between these two contract.
+     *
+     * @param asset Asset address
+     * @param amount Min amount of token to receive
+     */
+    function withdrawAssetFromActivePositionOnlyVault(
+        address asset,
+        uint256 amount
+    ) external onlyVault {
+        Position memory position = tokenIdToPosition[activeTokenId];
+        require(position.exists && position.liquidity > 0, "Liquidity error");
+
+        // Figure out liquidity to burn
+        (
+            uint128 liquidity,
+            uint256 minAmount0,
+            uint256 minAmount1
+        ) = _calculateLiquidityToWithdraw(position, asset, amount);
+
+        // Liquidiate active position
+        _decreasePositionLiquidity(
+            position.tokenId,
+            liquidity,
+            minAmount0,
+            minAmount1
+        );
```

Comment:
@shahthepro should we deposit the leftover assets to reserve strategies at this point by calling `_depositAll`?

Reply:
The funds (of the requested asset) removed from the active position are then transferred to the Vault. This function is also actually delegate called from another one in UniswapV3Startegy. So, we cannot call depositAll here. If we add that at the end of the withdraw on the other contract. It'll add up the gas costs of redeem. That's the reason I didn't wanna add it

Reply:
oh I see thanks for explanation